Add support for TP-Link smartbulbs

mjg59 opened the pull request:
This patchset adds support for TP-Link smartbulbs running the latest firmware. There appears to have been a slight modification to the protocol - the previously empty header is now the length of the request. It's possible that the emeter changes are firmware-version dependent rather than specific to the bulbs, so that may need to be refactored when we get more feedback.

Commit:
Create a SmartDevice parent class

Add a generic SmartDevice class that SmartPlug can inherit from, in
preparation for adding support for other device types.

Commit:
Add support for new-style protocol

Newer devices (including my LB130) seem to include the request length in
the previously empty message header, and ignore requests that lack it. They
also don't send an empty packet as the final part of a response, which can
lead to hangs. Add support for this, with luck not breaking existing devices
in the process.

Commit:
Add support for TP-Link smartbulbs

These bulbs use the same protocol as the smart plugs, but have additional
commands for controlling bulb-specific features. In addition, the bulbs
have their emeter under a different target and return responses that
include the energy unit in the key names.

mweinelt commented on pyHS100/__init__.py:
```diff
@@ -1,3 +1,3 @@
 from __future__ import absolute_import
 from __future__ import unicode_literals
-from pyHS100.pyHS100 import SmartPlug, TPLinkSmartHomeProtocol, SmartPlugException
+from pyHS100.pyHS100 import SmartBulb, SmartPlug, TPLinkSmartHomeProtocol, SmartPlugException
```
Ignore this.

rytilahti replied:
#31 will fix this.

Commit:
Fix tests

We now include the request length in the encrypted packet header, so strip
the header rather than assuming that it's just zeroes.

mjg59 commented on pyHS100/pyHS100.py:
```diff
@@ -37,37 +37,7 @@ class SmartPlugException(Exception):
-
-    ALL_FEATURES = (FEATURE_ENERGY_METER, FEATURE_TIMER)
-
+class SmartDevice(object):
```
I can't figure out what this complaint is - there appear to be two blank lines after this patch.

mweinelt replied:
Looks good to me as well, the SmartDevice class could use a docstring though.

mjg59 commented:
The Python 2 CI failure doesn't seem to be related to any of my changes?

mweinelt commented:
No, they're related to the "raise from" Syntax, which is not supported in Python 2. We'll have to use the implementation in six or future instead.

mjg59 commented:
Ok. Any other changes you'd like? I think the remaining hound issue is spurious.

mweinelt commented on pyHS100/tests/test_protocol.py:
```diff
@@ -10,6 +10,6 @@ class TestTPLinkSmartHomeProtocol(TestCase):
         encrypted = TPLinkSmartHomeProtocol.encrypt(d)
-        # encrypt appends nullbytes for the protocol sends
-        encrypted = encrypted.lstrip(b'\0')
+        # encypt adds a 4 byte header
```
typo, encrypt

mweinelt commented:
I'd like some feedback from @rytilahti and @kirichkov.

mweinelt commented on pyHS100/pyHS100.py:
```diff
@@ -534,3 +279,477 @@ def mac(self, mac):
+
+    @property
+    def has_emeter(self):
+        return True
```
Does the bulb expose a feature list in its sysinfo? If so, please check for emeter in that list.

mjg59 replied:
It doesn't, annoyingly. As far as I can tell all the bulbs have this feature.

rytilahti commented:
Unit-tests run fine here, examples.cli works with HS110. It'd probably make sense to split up pyHS100.py to SmartBulb, SmartPlug and SmartDevice, the file is getting big.

@mjg59 may we get also unit-tests for bulb functionality?
edit: please also add contents of sysinfo to unit-test file.

@trainman419 do you mind testing if these changes make HS105 work too?

rytilahti commented on pyHS100/pyHS100.py:
```diff
@@ -534,3 +279,477 @@ def mac(self, mac):
+        self.is_variable_color_temp = self.sys_info['is_variable_color_temp']
+        self.emeter_type = "smartlife.iot.common.emeter"
+        self.emeter_units = True
+
```
Define is_color, is_dimmable and is_variable_color_temp as properties, and skip accessing sysinfo here. Allows lazy initialization in, e.g., home assistant.

mjg59 replied:
That was my initial approach, but it seemed like it'd lead to more network activity in general.

rytilahti replied:
That's true, #26 changed the behaviour to disable caching of responses from the bulb so that every access to sys_info will trigger a request. I'm having second thoughts if it'd make more sense to revert that and make clients to adapt to #20.

kirichkov commented on pyHS100/pyHS100.py:
```diff
@@ -38,36 +38,7 @@ class SmartPlugException(Exception):
+class SmartDevice(object):
     def __init__(self, ip_address, protocol=None):
         """
         Create a new SmartPlug instance, identified through its IP address.
```
Change the comment to reflect the change in the class name, i.e. SmartDevice

kirichkov commented:
I also think it's high time we split up pyHS100.py into several smaller files.

Also I think this module needs a name change badly to properly reflect the devices it supports, but this should be low on the priority list.

Commit:
Add tests for bulbs

Not entirely comprehensive, but has pretty much the same level of testing
as plugs

rytilahti commented:
@mjg59 could you also please implement features in some form, even if it's not reading from "features" but based on some other keys? Those hound checks against fakes.py you can ignore, #31 will add a proper ignore when it gets merged.
edit: those can also be fixed, but I would not consider over-length lines a problem especially with fake test data..

InblEric commented:
Probably worth posting here as well but this PR did make my HS105 work.

mweinelt commented:
Huh, this merge was unfortunately a bit premature.

InblEric commented:
Did my comment trigger an automatic merge somehow? I just wanted to comment about it helping for the HS105 too.

kirichkov commented:
@InblEric no, don't worry about it! Merges are still manual here :-)

GadgetReactor commented:
Oops. Should I reverse this?

trainman419 commented:
I've confirmed that this works with my HS105 too!

mweinelt commented:
@GadgetReactor  Well, the code needs to be updated for PEP8 compliance, see @houndci-bot comments.

Also we didn't get the split into another module, this leaves `pyHS100.py` a 800 LOC monster. Should also be addressed imo.
